site-service/src/main/java/uk/ac/ox/ndph/mts/site_service/service/SiteServiceImpl.java

```diff
@@ -163,5 +200,124 @@ Site findRootSite() throws ResponseStatusException {
+            siteAddress.getPostcode());
+    }
+
+    @Override
+    public boolean filterUserSites(List<SiteDTO> sitesReturnObject,
+                                   String userRole, String accessPerm) {
+        try {
+            Objects.requireNonNull(sitesReturnObject, "sites can not be null");
+
+            //get unfiltered roleAssignments
+            List<RoleAssignmentDTO> roleAssnmts = new ArrayList<>(
+                practServClnt.getUserRoleAssignments(authService.getUserId(), authService.getAuthHeaders()));
+
+            //return 403 if RAs is empty
+            if (roleAssnmts.isEmpty()) {
+                logger.info("User with id {} has no role assignments and therefore is unauthorised.",
+                    authService.getUserId());
+                return false;
+            }
+
+            //only keep the assigned sites (or child assigned sites)
+            removeUnassignedSites(roleAssnmts, sitesReturnObject);
+
+            //If we dont have receive a role then return the assigned sites
+            if (userRole == null) {
+                return true;
+            }
+
+            //Remove RAs if they dont match the role
+            roleAssnmts.removeIf(ra -> !ra.getRoleId().equalsIgnoreCase(userRole));
+
+            //If RAs are empty return 403
+            if (roleAssnmts.isEmpty()) {
+                logger.info("User with id {} has no role assignments and therefore is unauthorised.",
+                    authService.getUserId());
+                return false;
+            }
+
+            //Filter sites by role
+            removeUnassignedSites(roleAssnmts, sitesReturnObject);
+
+            //Get the RAs with the given permission - if they are then empty return 403
+            List<RoleAssignmentDTO> roleAsstsWtPerm = getRoleAssmntsWtPerm(accessPerm, roleAssnmts);
+            if (roleAsstsWtPerm.isEmpty()) {
+                return false;
+            }
+
+            //Filter sites with perm based RAs
+            //check if user has site (or child site) assigned
+            removeUnassignedSites(roleAsstsWtPerm, sitesReturnObject);
+
+            //If sites are empty return 403 (false)
+            return !sitesReturnObject.isEmpty();
+        } catch (Exception e) {
+            return false;
+        }
+
+    }
+
+    private void removeUnassignedSites(List<RoleAssignmentDTO> roleAssignments, List<SiteDTO> sites) {
+        sites.removeIf(site ->
+            !siteUtil.getUserSites(sites, roleAssignments)
+                .contains(site.getSiteId()));
+    }
+
+    /**
+     * Validate if the role assignments have the required permission linked to them.
+     * @param reqPerm action required permission
+     * @param roleAssmnts user role assignments
+     * @return true if required permission is present in one of the roles
+     */
+    private List<RoleAssignmentDTO> getRoleAssmntsWtPerm(String reqPerm,
```

Comment:
Is getRoleAssmntsWtPerm the duplicated method that we discussed earlier?  Needed in both site-service and the authz component?   I am still not sure whether it's better to start an mts-utilities library (risking creeping back towards a monolith) or to relax about duplication.   Just recording our thoughts here for ref.